wallet-new/src/Cardano/Wallet/Kernel/Submission.hs

```diff
@@ -0,0 +1,431 @@
+    in ws & over (wsState . wssSchedule) (scheduleNext currentSlot)
+    where
+        scheduleNext :: Slot -> Schedule -> Schedule
+        scheduleNext currentSlot (Schedule s n) =
+            let txs     = map toEntry (pending ^. pendingTransactions . fromDb . to M.toList)
+                toEntry (txId, txAux) = ScheduledTx txId txAux (SubmissionCount 0)
+            in Schedule (IntMap.insertWith mappend (castSlot $ mapSlot succ currentSlot) txs s) n
+
+-- | Purges all the scheduled transactions in the given 'Slot'. This is
+-- automatically called by 'tick', and it's responsibility of @rho@ to
+-- correctly reschedule all the transaction.
+purgeSlot :: Slot -> Schedule -> Schedule
+purgeSlot slot (Schedule s n) = Schedule (IntMap.delete (castSlot slot) s) n
+
+-- Ready-to-use 'ResubmissionFunction's.
+
+-- | A 'RetryPolicy' is simply a function which instruct the 'Schedule' when
+-- to attempt resubmitting the given 'ScheduledTx' item.
+type RetryPolicy = SubmissionCount -> Slot -> Maybe Slot
+
+limited :: MaxRetries -> RetryPolicy -> RetryPolicy
+limited maxRetries retryPolicy = \retries currentSlot ->
+    case getSubmissionCount retries >= maxRetries of
+         True  -> Nothing
+         False -> retryPolicy retries currentSlot
+
+type Exponent   = Double
+type MaxRetries = Int
+
+--
+-- Stock retry policies inspired by the 'retry' package.
+--
+
+-- | Very simple policy which merely retries to resubmit the very next 'Slot',
+-- up until 'MaxRetries' attempts.
+constantRetry :: MaxRetries -> RetryPolicy
+constantRetry maxRetries =
+    limited maxRetries $ \_ currentSlot -> Just (mapSlot succ currentSlot)
+
+-- | An exponential backoff policy, parametric over a maximum number of
+-- 'MaxRetries' and an 'Exponent' for the backoff.
+exponentialBackoff :: MaxRetries -> Exponent -> RetryPolicy
+exponentialBackoff maxRetries exponent =
+    limited maxRetries $ \retries currentSlot ->
+        Just $ mapSlot (\s -> s + floor (exponent ^^ (getSubmissionCount $ retries))) currentSlot
+
+-- | A very customisable resubmitter which can be configured with different
+-- retry policies.
+defaultResubmitFunction :: forall m. Monad m
+                        => (Core.TxAux -> m Bool)
+                        -> RetryPolicy
+                        -> ResubmissionFunction m
+defaultResubmitFunction send retryPolicy currentSlot scheduled oldSchedule = do
+    foldlM updateFn (emptyPending, oldSchedule) scheduled
+    where
+        updateFn :: (Evicted, Schedule) -> ScheduledTx -> m (Evicted, Schedule)
+        updateFn (evicted, acc@(Schedule s nursery)) entry = do
+            let (ScheduledTx txId txAux retries) = entry
+                reschedule targetSlot =
+                    IntMap.insertWith mappend
+                                      (castSlot targetSlot)
+                                      [incSubmissionCount entry]
+                                      s
+            case retryPolicy retries currentSlot of
```

Comment:
The order here is correct (I _think_; this is the second time I'm writing this comment, I had the opposite conclusion first :) but it's confusing. We find a transaction that is due _now_, but instead of sending it, we _first_ check the retry policy to see if we should try again. That's.. odd. Not entirely sure what to suggest here, perhaps we should discuss it briefly. A comment or a different name or something may help. (Or maybe I'm just missing the point.)

Reply:
Yea, I see what you mean, and I suspect we might need to do a bit of rubber duck debugging to see what's make most sense -- I originally would have though that checking if we were meant to resubmit immediately would have avoided work, but I'm not concerned that, in case of particular retry policy, we might even end up in a situation where transactions _never attempted_ would be erroneously dismissed by a premature check.. 🤔